proposals/windows-mdm-commands-tables.md

```diff
@@ -0,0 +1,118 @@
+# Changes to Windows MDM command tables
+
+**Context** Commands are the mechanism we'll use to build features like
+
+- Deliver OS settings
+- Filter hosts by their "OS settings" status (pending, verified, verifying, failed)
+
+IF possible I would like us to have the same mental model for both platforms,
+where we have three tables with a clear separation of concerns:
+
+A. List of commands
+B. Queue
+C. List of command results
+
+Any command, doesn't matter how it was enqueued (via the UI/CLI or in response
+to other commands) should have an entry in `A`, plus entries with FK references
+in `B` when it's enqueued, and `C` when we get results.
+
+The [current schema](https://github.com/fleetdm/fleet/blob/defad14dd92ac1fb049552ec1564bceb50606e2d/server/datastore/mysql/migrations/tables/20230926082152_AddWindowsMDMTables.go)
+for Windows has two tables, with the following descriptions:
+
+### `windows_mdm_pending_commands`
+
+> The table windows_mdm_pending_commands contains the stateless commands enqueued into the MDM pending commands table from fleetctl or fleet Server.
+> These commands contain operations to be performed on a given device. 
+> They don't yet have the MessageID, SessionID, and CmdId information as this data gets populated during the response 
+> to a given MDM management request.
+
+### `windows_mdm_commands`
+
+> The windows_mdm_commands table represents the stateful commands that were actually sent over the wire. 
+> These commands include the MessageID, SessionID, and CmdId as they were sent to the device. 
+> They are stored in the commands table for error code tracking purposes. 
+> Additionally, this table can contain commands that were not part of the windows_mdm_pending_commands table and 
+> were generated by the Fleet server in response to an incoming MDM command.
+
+## Proposal
+
+I think it's fundamental we have:
+
+- A clear queue model
+- A single source of truth for commands
+- The status of a command at any point in time easily available
+- The raw XML we sent to the device
+- The raw XML we received as a response
+
+A rough version of the schema (many missing fields from the original) could be:
+
+```sql
+CREATE TABLE windows_mdm_commands (
+    -- managed and generated by Fleet
+    command_uuid    VARCHAR(127) NOT NULL,
+
+    -- raw XML
+    raw_command      MEDIUMTEXT   NOT NULL,
+
+    -- command_id specifies a unique command identifier for the SyncML message (Section 2.2.3.2 on MS-MDM spec)
+    command_id      VARCHAR(255) NOT NULL,
+
+    -- cmd_verb is e.g. Add, Exec, Get, etc.    
+    cmd_verb        VARCHAR(20)  NOT NULL,
```

Comment:
Is that still important to have as a standalone field, since it's part of the raw xml? If that's something we'll need to display somewhere or filter by, then yes for sure.

Also, I think we'd need something like what was called `system_origin` in the previous schema, to identify it as fleet-initiated (well, I'm not sure we absolutely need to differentiate those, but it seems like a relevant information that we might want going forward, and if we YAGNI it, it might be tricky to backport).

Reply:
And depending on what information we want to display in `fleetctl get mdm-command-results`, I think we'd want to store the target URL (what was called `system_uri`) as a distinct field to simplify this (although this one is less obvious as a requirement, if we wanted to retrieve it from the raw xml we'd have to re-parse the xml before returning the information).

Reply:
yes, I agree that `system_origin` would be helpful to determine which commands were generated by the Fleet server or by some user input (admin following gitops flow)

On the other hand, having `system_uri` would require parsing the provided raw XML to extract the target OMA-URI property. If this is not needed for the profile feature, we might avoid tracking it

Reply:
Yeah I don't mind not storing `system_uri` separately for now if we don't think we'll need it. My understanding was that we do need it for the `fleetctl mdm get-command-result` payload, but I may be wrong.